[R-package] add a tree plotting function

fboudry opened the pull request:
Feature requested in https://github.com/microsoft/LightGBM/issues/1222

Added a R function to plot trees.
Basically used the code posted in https://github.com/microsoft/LightGBM/issues/1222 by @SpeckledJim2 and followed the given instruction. 

Plot should look like this:
![Screenshot_20250315_161929](https://github.com/user-attachments/assets/2d4f1eda-8892-4a4d-ba61-f97e89e5dd6f)

fboudry committed:
Added lgb.plot.tree function.
Added DiagrammeR as suggested in DESCRIPTION
Added lgb.plot.tree in _pkgdown.yml
Roxygenized.

fboudry committed:
Added tests.

fboudry commented:
@microsoft-github-policy-service agree

jameslamb commented on R-package/R/lgb.plot.tree.R:
```diff
@@ -0,0 +1,184 @@
+        nodes_df = nodes,
+        edges_df = edges,
+        attr_theme = NULL
+    ) %>%
```
In this project, by convention we:

* do not use the `%>%` operator
* use comma-first style everywhere

Please update this code and all the other code you're adding to follow that. Keeping all of the code looking the same across the codebase helps us to develop and review changes.

jameslamb replied:
`xgboost`'s implementation of similar functionality might be useful as a reference. See https://github.com/dmlc/xgboost/blob/e988b7cf1515b08ad0f949c26beb043ce0b33fe8/R-package/R/xgb.plot.tree.R#L159-L181

jameslamb commented on R-package/tests/testthat/test_lgb.plot.tree.R:
```diff
@@ -0,0 +1,59 @@
+test_that("lgb.plot.tree works as expected"){
```
Please also add tests for the other types of machine learning tasks LightGBM can be used for:

* binary classification
* multiclass classification (where, please note, there are `num_classes` trees produced per iteration)
* learning-to-rank

And for the following model situations:

* uses categorical features

These are all cases that could affect the code as written... for example, categorical features have different splitting rules.

jameslamb commented on R-package/R/lgb.plot.tree.R:
```diff
@@ -0,0 +1,184 @@
+    dt <- dt[tree_index == tree, ]
+    # change the column names to shorter more diagram friendly versions
+    data.table::setnames(dt, old = c("tree_index", "split_feature", "threshold", "split_gain"), new = c("Tree", "Feature", "Split", "Gain"))
+    dt[, Value := 0.0]
+    dt[, Value := leaf_value]
```
```suggestion
    dt[, Value := leaf_value]
```

I don't understand this... what's the purpose of setting all rows to `0.0` and then immediately overwriting them? It seems to me that the `0.0` could probably be removed.

jameslamb commented on R-package/R/lgb.plot.tree.R:
```diff
@@ -0,0 +1,184 @@
+    data.table::setnames(dt, old = c("tree_index", "split_feature", "threshold", "split_gain"), new = c("Tree", "Feature", "Split", "Gain"))
+    dt[, Value := 0.0]
+    dt[, Value := leaf_value]
+    dt[is.na(Value), Value := internal_value]
+    dt[is.na(Gain), Gain := leaf_value]
+    dt[is.na(Feature), Feature := "Leaf"]
+    dt[, Cover := internal_count][Feature == "Leaf", Cover := leaf_count]
+    dt[, c("leaf_count", "internal_count", "leaf_value", "internal_value") := NULL]
+    dt[, Node := split_index]
+    max_node <- max(dt[["Node"]], na.rm = TRUE)
+    dt[is.na(Node), Node := max_node + leaf_index + 1]
+    dt[, ID := paste(Tree, Node, sep = "-")]
+    dt[, c("depth", "leaf_index") := NULL]
+    dt[, parent := node_parent][is.na(parent), parent := leaf_parent]
+    dt[, c("node_parent", "leaf_parent", "split_index") := NULL]
+    dt[, Yes := dt$ID[match(dt$Node, dt$parent)]]
+    dt <- dt[nrow(dt):1, ]
+    dt[, No := dt$ID[match(dt$Node, dt$parent)]]
+    # which way do the NA's go (this path will get a thicker arrow)
+    # for categorical features, NA gets put into the zero group
+    dt[default_left == TRUE, Missing := Yes]
+    dt[default_left == FALSE, Missing := No]
```
Can you please add some comments to make it a bit easier to understand what's happening in this wall of code? It's very difficult to read (at least for me) as currently written).

jameslamb commented on R-package/tests/testthat/test_lgb.plot.tree.R:
```diff
@@ -0,0 +1,59 @@
+  # trees start from 0 in lgb.model.dt.tree
+  tree_table <- lgb.model.dt.tree(model)
+  expect_error({
+    lgb.plot.tree(model, 999)TRUE
```
```suggestion
    lgb.plot.tree(model, 999)
```

This looks like it was included accidentally?

jameslamb reviewed (changes requested):
Thanks for your interest in LightGBM.

As I mentioned in the discussion on #1222, I'm supportive of trying to add something like this ([especially since xgboost has it as well](https://github.com/dmlc/xgboost/blob/e988b7cf1515b08ad0f949c26beb043ce0b33fe8/R-package/R/xgb.plot.tree.R)).

But I hope you'll see from the first round of suggestions I left here... significant work remains before I'd support merging this change into the package. If you are willing to work with us on this and go through multiple rounds of reviews and suggestions, we'd be grateful for the help! But if you don't have the time/interest to get this ready for inclusion in the package, please let me know and we'll close this PR and leave #1222 open for someone else to pick up.

fboudry committed:
Added review suggestions.
DiagrammeR in CI.
Error messages.
Default parameters.
Changed tests.

fboudry committed:
Updated tests. (based on R-package/tests/testthat/test_lgb.model.dt.tree.R)
Now tests regressions, binary, multiclass classification and ranks.

fboudry committed:
Corrected error (missing comma) in the selected tree check (L66).
Commented code.

fboudry committed:
Corrected tests.
Added a warning to functions and shorter stop message to make tests work.

fboudry committed:
Added DiagrammeR in CI, github and Azure workflows as well as in the README.

fboudry commented:
Thanks for the review @jameslamb, helped me a lot!
I think I've made all suggested changes, however the checks are failing on the "R CMD check". When run locally I also have an error about the Description file (missing maintainer field). I don't really know what I'm supposed to do with that/how you want to handle it in this package so I'll take any suggestion!

fboudry commented on R-package/DESCRIPTION:
```diff
@@ -49,7 +49,8 @@ Suggests:
     RhpcBLASctl,
-    testthat
+    testthat,
+    DiagrammeR
```
Using `processx` isn't enough to find all places in the CI files, I suggest using `knitr` instead, that's where I added `DiagrammeR`.

jameslamb replied:
ah ok sure, that works. Thanks.

jameslamb commented on R-package/DESCRIPTION:
```diff
@@ -45,11 +45,12 @@ NeedsCompilation: yes
     processx,
     RhpcBLASctl,
-    testthat
+    testthat,
```
```suggestion
    testthat
```

@fboudry I suspect this trailing comma is why `R CMD check` is failing. I don't think trailing commas in lists are supported in `DESCRIPTION` (just as they are not supported in R).

Let's remove this.

jameslamb committed:
Update R-package/DESCRIPTION

jameslamb commented on R-package/R/lgb.plot.tree.R:
```diff
@@ -0,0 +1,204 @@
+#' }
+#'
+#' @export
+lgb.plot.tree <- function(model = NULL, tree = NULL, rules = NULL) {
```
```suggestion
lgb.plot.tree <- function(model, tree, rules = NULL) {
```

I can't think of any situation where it would be ok for `model` or `tree` to be `NULL`, can you?

If not, let's please require callers to provide values explicitly.

jameslamb commented on R-package/R/lgb.plot.tree.R:
```diff
@@ -0,0 +1,204 @@
+            call. = FALSE
+        )
+    }
+    # tree must be numeric
+    if (!inherits(tree, "numeric")) {
+        stop("lgb.plot.tree: Has to be an integer numeric")
+    }
+    # tree must be integer
+    if (tree %% 1 != 0) {
+        stop("lgb.plot.tree: Has to be an integer numeric")
+    }
```
```suggestion
    # tree must be numeric
    tree <- as.integer(tree)
    if (length(tree) != 1L || tree < 1L) {
        stop(sprintf("lgb.plot.tree: 'tree' must be a single, positive integer.)
    }
```

Let's combine these, and make it clear *what* has to be an integer.

jameslamb commented on R-package/R/lgb.plot.tree.R:
```diff
@@ -0,0 +1,204 @@
+    modelDT <- lgb.model.dt.tree(model)
+    # check that tree is less than or equal to the maximum tree index in the model
+    if (tree > max(modelDT$tree_index) || tree < 1) {
+        warning("lgb.plot.tree: Value of 'tree' should be between 1 and the total number of trees in the model (", max(modelDT$tree_index), "). Got: ", tree, ".")
+        stop("lgb.plot.tree: Invalid tree number")
```
```suggestion
        stop("lgb.plot.tree: Value of 'tree' should be between 1 and the total number of trees in the model (", max(modelDT$tree_index), "). Got: ", tree, ".")
```

What's the reason for having all of the information in a `warning()` and then immediately raising an error after? If there isn't a specific reason, then let's please combine these for simplicity and to make the logs easier for users to understand.

jameslamb commented on R-package/R/lgb.plot.tree.R:
```diff
@@ -0,0 +1,204 @@
+    return(invisible(NULL))
+}
+
+.levels.to.names <- function(x, feature_name, rules) {
```
```suggestion
.levels_to_names <- function(x, feature_name, rules) {
```

Please avoid using `.` in any of these private functions' names.

jameslamb commented on R-package/R/lgb.plot.tree.R:
```diff
@@ -0,0 +1,204 @@
+#' @description The \code{lgb.plot.tree} function creates a DiagrammeR plot of a single LightGBM tree.
+#' @param model a \code{lgb.Booster} object.
+#' @param tree an integer specifying the tree to plot. This is 1-based, so e.g. a value of '7' means 'the 7th tree' (tree_index=6 in LightGBM's underlying representation).
+#' @param rules a list of rules to replace the split values with feature levels.
```
I'm not totally convinced about this idea... it should be possible to recover the feature names from the model directly.

But before you remove this... can you please expand this doc and add examples and tests showing what this would look like? Right now, it's hard for me to understand what the content of `rules` is supposed to be.

jameslamb commented on R-package/R/lgb.plot.tree.R:
```diff
@@ -0,0 +1,204 @@
+    # filter modelDT to just the rows for the selected tree
+    modelDT <- modelDT[tree_index == tree, ]
+    # change the column names to shorter more diagram friendly versions
+    data.table::setnames(modelDT
+    , old = c("tree_index", "split_feature", "threshold", "split_gain")
+    , new = c("Tree", "Feature", "Split", "Gain"))
```
```suggestion
    data.table::setnames(
        modelDT
        , old = c("tree_index", "split_feature", "threshold", "split_gain")
        , new = c("Tree", "Feature", "Split", "Gain")
    )
```

Please, follow the style the rest of the project uses. I suspect that the linting configuration here would have caught this (not sure, as I haven't run it myself and it failed in CI for other unrelated reasons).

From this point forward, before you push a commit please run the R-code linting and fix any issues it reports.

From the root of the repo:

```shell
Rscript ./.ci/lint-r-code.R ./R-package
```

fboudry replied:
We changed this (it is definitely easier to read) but the linter did not showed the first line as having to be changed.

jameslamb reviewed (changes requested):
Thanks. I left a few more comments for your consideration. To be clear, I still haven't very-thoroughly reviewed so this is not a comprehensive list... these are just quick things I noticed in the few minutes I had to review.

In addition to those... it'd be helpful if you could include some screenshots of what the plots look like, in the description of the PR. That'll really help me understand what the goal is here, without needing to run this code myself.

Commit:
Fix linting, amend test wording

Commit:
Fix linting, make changes based on LightGBM maintainer feedback, better Help examples including categorical features

fboudry committed:
Adressing [PR review](https://github.com/microsoft/LightGBM/pull/6729#pullrequestreview-2526665067).

- `model` and `tree` parameters made mandatory
- fixed linting
- combined tree type tests

fboudry commented:
Hi @jameslamb ! It took us some time but we worked on this with @SpeckledJim2. I think we've addressed all of your comments/change requests. If you have any more suggestions or requests we'll look into it! Thanks.

jameslamb commented:
> I think we've addressed all of your comments/change requests.

Thanks. I'll try to look some time this week.

To reduce the effort it'll take us to review, please respond to each of the open threads with a link to a commit or a brief explanation of how you addressed the feedback.

jameslamb commented:
@fboudry sorry for the long delay. If you could please update this with the lastest changes on `master` and respond to all the open threads (marking any that are complete "resolved" along with a comment), I'd be happy to review this again.
